### ISSUE TITLE
DynamicScale goes to infinity

### ISSUE THREAD

**Issue by Gurvan** [google/flax#2547]
### System information
- OS Platform and Distribution: Linux 5.15.74-3-MANJARO
- Flax, jax, jaxlib versions: flax 0.6.0, jax 0.3.21, jaxlib 0.3.20+cuda11.cudnn82
- Python version: 3.10.4
- GPU/TPU model and memory: NVIDIA RTX 2060 Super 8GB
- CUDA version (if applicable): cuda11, cudnn82

### Problem you have encountered:
I adapted the gradient scaling code of the imagenet example in order to use fp16 training for my model. During training, the `scale` parameter of `DynamicScaling` became `inf`, because it kept doubling every 2000 steps without the gradients ever becoming `nan`. Once the `scale` hit `inf` it never went down, since dividing `inf` by any number still returns `inf`.

### What you expected to happen:
Either have a `maximum_scale` parameter or add a check to see if `scale` is finite.
I think it's a one line change in the `DynamicScale` code, so I can make a pull request if you agree.

### Steps to reproduce:
```py
import jax.numpy as jnp
from flax.training import dynamic_scale

ds = dynamic_scale.DynamicScale(growth_factor=10.0, growth_interval=1)
x = jnp.float32(1)

for _ in range(70):
    ds, is_fin, aux, grad = ds.value_and_grad(lambda p: p**2)(x)
    x = x - 0.1 * grad
    print(is_fin, ds.scale, x, grad)
```

### Code:
Could check here ([dynamic scale code](https://github.com/google/flax/blob/6b80cbb239b07e370fd0af6655b39ef40de061c0/flax/training/dynamic_scale.py#L139)) if `fin_scale` is finite.

**Comment by jheek** [comment]
I think we should add an upper bound indeed, are you sure you are really computing in fp16? I'm a bit surprised that this is something you can actually hit in practice

**Comment by Gurvan** [comment]
The model parameters are float16, the model inputs are int16 and I accumulate the loss in float32. Now that I think about it, since the loss is accumulated as float32, the grads are float32 (which probably explains why they don't become NaNs). Does it mean that all the backward pass is done in float32?

**Comment by Gurvan** [comment]
After checking more extensively, I figured out that part of my model inputs were actually float32. After fixing it the training is about twice as fast, so I wasn't computing in fp16 after all! Thank you for pointing it out.

In any case, it still might be useful to make sure that the scale never goes to infinity.
